# Contents
1. Question 3
2. Question 4
3. Question 5
4. Question 6

# Question 3

In [1]:
#Import libraries and dataset
import pandas as pd
import numpy as np
import os

In [2]:
path = r'C:\Users\Owner\OneDrive\Desktop\Data Analytics\Achievement 4\11-2022 Instacart Basket Analysis'

In [3]:
customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

In [4]:
customers.head(15)

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
0     26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1     33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2     65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3    125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4    130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
5    133128    Cynthia     Noble  Female        Kentucky   43    1/1/2017   
6    152052      Chris    Walton    Male         Montana   20    1/1/2017   
7    168851     Joseph   Hickman    Male  South Carolina   30    1/1/2017   
8     69965     Jeremy      Vang    Male           Texas   47    1/1/2017   
9     82820      Shawn     Chung    Male        Virginia   26    1/1/2017   
10    26441     Gloria  Stafford  Female          Nevada   19    1/1/2017   
11    33835      Roger     Casey    Male        Nebraska   45    1/1/2017   
12    75032    Cynthia  Mcdaniel  Female         Georgia   75    1/1/2017   
13       51  Stephanie    Jensen  Female         Wyoming   37    1/1/2017   
14      516      Peter      Hunt    Male        Colorado   51    1/1/2017   

    n_dependants                        fam_status  income  
0              3                           married  165665  
1              0                            single   59285  
2              2                           married   99568  
3              0                            single   42049  
4              1                           married   40374  
5              2                           married   49643  
6              0                            single   61746  
7              0                            single   63712  
8              1                           married  162432  
9              2                           married   32072  
10             2  living with parents and siblings   43443  
11             2                           married   47988  
12             1                           married   57835  
13             1                           married   71190  
14             2                           married  146559

In [5]:
customers.shape

(206209, 10)

# Question 4

In [10]:
#Wrangling Data-Dropping n_dependants from dataframe
customers1 = customers.drop(columns = ['n_dependants'])

In [6]:
customers1.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

  fam_status  income  
0    married  165665  
1     single   59285  
2    married   99568  
3     single   42049  
4    married   40374

In [7]:
#Renaming surnam, STATE, and fam_status
customers1.rename(columns = {'Surnam' : 'Last Name'}, inplace = True)

In [8]:
customers1.rename(columns = {'STATE' : 'State'}, inplace = True)

In [9]:
customers1.rename(columns = {'fam_status' : 'family_status'}, inplace = True)

In [10]:
customers1.head()

user_id First Name Last Name  Gender       State  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

  family_status  income  
0       married  165665  
1        single   59285  
2       married   99568  
3        single   42049  
4       married   40374

# Question 5

In [11]:
#Check for missing values
customers1.isnull().sum()

user_id              0
First Name       11259
Last Name            0
Gender               0
State                0
Age                  0
date_joined          0
family_status        0
income               0
dtype: int64

In [12]:
df_nan = customers1[customers1['First Name'].isnull() ==  True]

In [13]:
df_nan

user_id First Name Last Name  Gender         State  Age date_joined  \
53        76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73        13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82        89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99        96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105       29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...         ...        ...       ...     ...           ...  ...         ...   
206038   121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044   200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090   167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162   187532        NaN     Floyd  Female    California   39    4/1/2020   
206171   116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

       family_status  income  
53           married   41709  
73            single   82518  
82           married  117099  
99           married  155673  
105          married  151819  
...              ...     ...  
206038       married   87783  
206044       married  108488  
206090       married   45275  
206162        single   56325  
206171       married   59222  

[11259 rows x 9 columns]

There are 11,259 missing entries, but they are all first names. This would have to be corrected at the source as it is valuable customer information. Although, it wouldn't be needed for anything in particular but to do things like loyalty programs. The rest of the data is intact, so I would keep the missing values.

In [14]:
#Finding duplicates
df_dups = customers1[customers1.duplicated()]

In [15]:
df_dups

Empty DataFrame
Columns: [user_id, First Name, Last Name, Gender, State, Age, date_joined, family_status, income]
Index: []

There are no duplicates.

In [16]:
#Finding mixed data types
for col in customers1.columns.tolist():
    weird = (customers1[[col]].applymap(type) != customers1[[col]].iloc
            [0].apply(type)).any(axis = 1)
    if len (customers1[weird]) > 0:
        print (col)

First Name


In [17]:
customers1['First Name'] = customers1['First Name'].astype('str')

In [18]:
#Finding mixed data types
for col in customers1.columns.tolist():
    weird = (customers1[[col]].applymap(type) != customers1[[col]].iloc
            [0].apply(type)).any(axis = 1)
    if len (customers1[weird]) > 0:
        print (col)

# Question 6

In [6]:
#Importing ords_prods_merged
ords_prods_merged = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged2.pkl'))

In [20]:
ords_prods_merged.head()

Unnamed: 0_x  order_id  customer_id  order_number  orders_day_of_week  \
0             0   2539329            1             1                   2   
1             1   2398795            1             2                   3   
2             2    473747            1             3                   3   
3             3   2254736            1             4                   4   
4             4    431534            1             5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered  ...  _merge    price_range_loc    busiest_days  \
0          0  ...    both  Mid-range product  Regularly busy   
1          1  ...    both  Mid-range product    Slowest days   
2          1  ...    both  Mid-range product    Slowest days   
3          1  ...    both  Mid-range product    Slowest days   
4          1  ...    both  Mid-range product    Slowest days   

   busiest_period_of_day  max_order  loyalty_flag average_price spending_flag  \
0         Average orders         10  New customer          14.0  High spender   
1         Average orders         10  New customer          14.0  High spender   
2            Most orders         10  New customer          14.0  High spender   
3         Average orders         10  New customer          14.0  High spender   
4            Most orders         10  New customer          14.0  High spender   

  median_days_since_prior_order         frequency_flag  
0                          30.0  non-frequent customer  
1                          30.0  non-frequent customer  
2                          30.0  non-frequent customer  
3                          30.0  non-frequent customer  
4                          30.0  non-frequent customer  

[5 rows x 25 columns]

In [21]:
customers1.head()

user_id First Name Last Name  Gender       State  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

  family_status  income  
0       married  165665  
1        single   59285  
2       married   99568  
3        single   42049  
4       married   40374

In [15]:
ords_prods_merged.rename(columns = {'customer_id' : 'user_id'}, inplace = True)

In [16]:
ords_prods_merge = ords_prods_merged.drop(columns = ['_merge'])

MemoryError: Unable to allocate 3.14 GiB for an array with shape (13, 32404859) and data type int64

In [14]:
#Merging ords_prods_merge w/ customers1
df_cust_ords_prods_merged = ords_prods_merge.merge(customers1, on = 'user_id', indicator = True)

KeyError: 'user_id'

In [25]:
df_cust_ords_prods_merged.head()

Unnamed: 0_x  order_id  user_id  order_number  orders_day_of_week  \
0             0   2539329        1             1                   2   
1             1   2398795        1             2                   3   
2             2    473747        1             3                   3   
3             3   2254736        1             4                   4   
4             4    431534        1             5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered  ...         frequency_flag First Name  Last Name  Gender  \
0          0  ...  non-frequent customer      Linda     Nguyen  Female   
1          1  ...  non-frequent customer      Linda     Nguyen  Female   
2          1  ...  non-frequent customer      Linda     Nguyen  Female   
3          1  ...  non-frequent customer      Linda     Nguyen  Female   
4          1  ...  non-frequent customer      Linda     Nguyen  Female   

     State Age date_joined family_status  income _merge  
0  Alabama  31   2/17/2019       married   40423   both  
1  Alabama  31   2/17/2019       married   40423   both  
2  Alabama  31   2/17/2019       married   40423   both  
3  Alabama  31   2/17/2019       married   40423   both  
4  Alabama  31   2/17/2019       married   40423   both  

[5 rows x 33 columns]

In [26]:
df_cust_ords_prods_merged['_merge'].value_counts()

both          32404859
left_only            0
right_only           0
Name: _merge, dtype: int64

In [28]:
df_cust_ords_prods_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_cust_merged.pkl'))